rust/connlib/libs/gateway/src/lib.rs

```diff
@@ -1,22 +1,245 @@
+                    }
+                }
+            });
+
+            // Used to send internal messages
+            let control_signaler = ControlSignaler { control_signal: connection.sender_with_topic("gateway".to_owned()) };
+            let tunnel = fatal_error!(
+                Tunnel::new(private_key, control_signaler.clone(), callbacks.clone()).await,
+                runtime_stopper,
+                &callbacks
+            );
+
+            let mut control_plane = ControlPlane {
+                tunnel: Arc::new(tunnel),
+                control_signaler,
+            };
+
+            tokio::spawn(async move {
+                let mut interval = tokio::time::interval(Duration::from_secs(10));
+                loop {
+                    tokio::select! {
+                        Some((msg, _)) = control_plane_receiver.recv() => {
+                            match msg {
+                                Ok(msg) => control_plane.handle_message(msg).await?,
+                                Err(_msg_reply) => todo!(),
+                            }
+                        },
+                        _ = interval.tick() => control_plane.stats_event().await,
+                        else => break
+                    }
+                }
+
+                Result::Ok(())
+            });
+
+            tokio::spawn(async move {
+                let mut exponential_backoff = ExponentialBackoffBuilder::default()
+                    .with_max_elapsed_time(None)
+                    .build();
+                loop {
+                    // `connection.start` calls the callback only after connecting
+                    tracing::debug!("Attempting connection to portal...");
+                    let result = connection.start(vec!["gateway".to_owned()], || exponential_backoff.reset()).await;
+                    tracing::warn!("Disconnected from the portal");
+                    if let Err(e) = &result {
+                        tracing::warn!(error = ?e, "Portal connection error");
+                    }
+                    if let Some(t) = exponential_backoff.next_backoff() {
+                        tracing::warn!("Error connecting to portal, retrying in {} seconds", t.as_secs());
+                        let _ = callbacks.on_error(&result.err().unwrap_or(Error::PortalConnectionError(tokio_tungstenite::tungstenite::Error::ConnectionClosed)));
+                        tokio::time::sleep(t).await;
+                    } else {
+                        tracing::error!("Connection to portal failed, giving up");
+                        fatal_error!(
+                            result.and(Err(Error::PortalConnectionError(tokio_tungstenite::tungstenite::Error::ConnectionClosed))),
+                            runtime_stopper,
+                            &callbacks
+                        );
+                    }
+                }
+
```

Comment:
For the gateway, I reckon we should actually do this outside of `Session` in `main.rs`.

Reply:
Given these changes, we don't even need a `Session` for the gateway

Reply:
Are you suggesting to inline everything into an `async fn main`?

Reply:
To keep the PR manageable, should we postpone this to a later PR?